libraries/AP_Scripting/lua_repl.cpp

```diff
@@ -103,29 +103,54 @@ int lua_scripts::pushline(lua_State *L, int firstline) {
     // send prompt to the user
     terminal_print(get_prompt(L, firstline));
 
-    while (terminal.session) {
-        // reseek to where we need input from, as invalid reads could have done weird stuff, and we want to start from the last valid input
-        int input_fd = AP::FS().open(REPL_IN, O_RDONLY);
-        if (input_fd != -1) {
-            AP::FS().lseek(input_fd, terminal.input_offset, SEEK_SET);
-            ssize_t read_bytes = AP::FS().read(input_fd, buffer, ARRAY_SIZE(buffer) - 1);
-            AP::FS().close(input_fd);
-            if (read_bytes > 0) {
-                // locate the first newline
-                char * newline_chr = strchr(buffer, '\n');
-                if (newline_chr != NULL) {
-                    newline_chr[0] = '\0';
-                    // only advance to the newline
-                    l = strlen(buffer);
-                    terminal.input_offset += l + 1;
-                    break;
+    bool got_line = false;
+    while (terminal.session && !got_line) {
+        switch (terminal.type) {
+            case AP_Scripting::relp_type::REPL_FILE: {
+            // reseek to where we need input from, as invalid reads could have done weird stuff, and we want to start from the last valid input
+            int input_fd = AP::FS().open(REPL_IN, O_RDONLY);
+            if (input_fd != -1) {
+                AP::FS().lseek(input_fd, terminal.input_offset, SEEK_SET);
+                ssize_t read_bytes = AP::FS().read(input_fd, buffer, ARRAY_SIZE(buffer) - 1);
+                AP::FS().close(input_fd);
+                if (read_bytes > 0) {
+                    // locate the first newline
+                    char * newline_chr = strchr(buffer, '\n');
+                    if (newline_chr != NULL) {
+                        newline_chr[0] = '\0';
+                        // only advance to the newline
+                        l = strlen(buffer);
+                        terminal.input_offset += l + 1;
+                        got_line = true;
+                    }
                 }
             }
+            // wait for any input
+            hal.scheduler->delay(100);
+            break;
+            }
+
+            case AP_Scripting::relp_type::REPL_SERIAL: {
+            // can only read a byte at a time as we have to look out for the newline
+            if (terminal.uart == nullptr) {
+                terminal.session = false;
+                return 0;
+            } else if (terminal.uart->available()) {
```

Comment:
This should really be reading in upto the number of available bytes at the start here. As a general best practice we want to avoid any unbounded loops reading from a UART or any other off system device without a sleep somewhere to yield to any tasks at equal or lower priority. At the moment this is mostly safe because scripting is considered very low priority, but if we ever add anything at the same or lower priority it could start to become a problem.